docs/design/2020-06-24-placement-rules-in-sql.md

```diff
@@ -0,0 +1,391 @@
+
+Parameter `label` should be a string and in such a format: `{+|-}key=value,...`. Prefix `+` indicates that data can only be placed on those TiKV nodes whose labels contain such labels, and `-` indicates that data can’t be placed on those TiKV nodes whose labels contain such labels. `key` here refers to the label name, and `value` is the label value. `key` should be defined in TiKV labels.
+
+For example, `+zone=sh,+zone=bj` indicates to place data only in `sh` and `bj` zone. Also, `zone` should have already be defined in TiKV configuration. For example:
+
+```sql
+[server]
+labels = "zone=<zone>,rack=<rack>,host=<host>"
+```
+
+Label constraints can be implemented by defining `label_constraints` field in placement rules. `+` and `-` correspond to property `op`. Specifically, `+` is equivalent to `In` and `-` is equivalent to `NotIn`.
+
+For example, `+zone=sh,+zone=bj` is equivalent to:
+
+```
+"label_constraints": [
+	{"key": "zone", "op": "in", "values": ["sh", “bj”]}
+]
+```
+
+### Role configuration
+
+`ROLE` in the statement defines the role of replicas. Newly added replicas are in such roles.
+
+`ROLE` is equivalent to property `role` in placement rules. it can be one of such values: leader, follower, learner, voter.
+
+`ROLE` must be specified in the statement.
+
+### Count configuration
+
+`COUNT` in the statement defines replicas count in current rule, it’s equivalent to property `count` in placement rules.
+
+`COUNT` must be specified in the statement.
+
+### Key range configuration
+
+Affected key range must also be defined in placement rules. Now that `table_name` is specified in the ALTER statement, key range can be inferred.
+
+Typically, key format is in such a format: `t_{table_id}_r_{pk_value}`, where `pk_value` may be `_tidb_rowid` in some cases. `table_id` can be inferred from `table_name`, thus key range can be `t_{table_id}_` to `t_{table_id+1}_`.
+
+### Database placement
+
+Defining placement of databases simplifies the procedures when there are very many tables. For example, in a multi-tenant scenarios, each user may have a single database. The dataset in one database is relatively small, and it’s rare to query toward multiple databases. In this case, a whole database can be placed in a single region to reduce multi-region latency.
+
+Placement of databases can be also defined through ALTER statements:
+
+```sql
+ALTER {DATABASE | SCHEMA} schema_name
+	{ADD | ALTER} PLACEMENT LABEL=label ROLE=role COUNT=count,
+	...
+```
+
+This statement indicates TiDB to add new replicas or alter placement for one database, including all tables in it.
+
+Since key range is not successive in one database, each table in the database should correspond to a placement rule, so there may be many placement rules.
+
+Creating or dropping a table should also affect the placement rules.
```

Comment:
Does it mean that if I add a placement rule for database, all tables (data, partition, index) created later for this database will automatically apply that rule?

Reply:
All tables in the database, including those created later.
This also applies for tables and partitions. For example, newly added partitions should follow the rules defined on the table it belongs to.